database/db.py

```diff
@@ -115,50 +71,186 @@ def __init__(self, username: str):
         super().__init__(f'user with username {username!r} was not found.')
 
 
-class FriendRequestNotFoundException(FailureException):
+class FriendRequestNotFoundException(DatabaseFailureException):
     """An exception to be raised if a friend request cannot be found."""
 
     def __init__(self, username, friend_username):
         super().__init__(f'No friend request found from {friend_username} on {username}\'s list.')
 
 
+@dataclass
+class DBTransaction:
+    statements: list[str] = field(default_factory=list)
+    was_committed: bool = False
+    was_rolled_back: bool = False
+
+    def add_statement(self, statement: str) -> None:
+        if self.was_committed:
+            raise TransactionClosedException("Cannot add a statement to a transaction that has been committed")
+
+        if self.was_rolled_back:
+            raise TransactionClosedException("Cannot add a statement to a transaction that has been rolled back")
+
+        self.statements.append(statement)
+
+    @property
+    def is_closed(self):
+        return self.was_committed or self.was_rolled_back
+
+
+class DBContext(ABC):
+
+    @property
+    @abstractmethod
+    def fetched(self) -> List[tuple]:
+        """Any fetched Data."""
+        ...
+
+    @property
+    @abstractmethod
+    def was_successful(self) -> bool:
+        """True if the transaction was committed successfully."""
+        ...
+
+    @abstractmethod
+    def execute(self, statement) -> None:
+        """Execute given statement on the database."""
+        ...
+
+    @abstractmethod
+    def commit(self) -> None:
+        """Commits the current transaction."""
+        ...
+
+    @abstractmethod
+    def fetchall(self) -> List[tuple]:
+        """Gets all rows of a query result, stores them in the context field `fetched` and also returned."""
+        ...
+
+
+class DBContextManager(ABC):
+
+    @abstractmethod
+    def __enter__(self) -> DBContext: ...
+
+    @abstractmethod
+    def __exit__(self, exc_type, exc_val, exc_tb) -> bool: ...
+
+
+class QueryBuilder(ABC):
```

Comment:
I created an abstract base class QueryBuilder, so regardless of what DB implementation we use we just have to extend this base class to create a new query builder for that implementation.